src/subhalo_merge.cpp

```diff
@@ -40,37 +36,51 @@ void SubHelper_t::BuildPosition(const Subhalo_t &sub)
     return;
   }
 
-  HBTInt NumPart = sub.Nbound;
-  if (NumPart > NumPartCoreMax)
-    NumPart = NumPartCoreMax;
-
-  HBTInt i, j;
+  // Initalize variables used to accumulate position, mass etc
+  HBTInt NumPart = 0;
   double sx[3], sx2[3], origin[3], msum;
-
   sx[0] = sx[1] = sx[2] = 0.;
   sx2[0] = sx2[1] = sx2[2] = 0.;
   msum = 0.;
+
+  // Use first particle as reference point for box wrap
   if (HBTConfig.PeriodicBoundaryOn)
-    for (j = 0; j < 3; j++)
+    for (int j = 0; j < 3; j++)
       origin[j] = sub.Particles[0].ComovingPosition[j];
+  
+  // Might need to make two passes through the particles
+  for(int pass_nr=0; pass_nr<2; pass_nr+=1) {
 
-  for (i = 0; i < NumPart; i++)
-  {
-    HBTReal m = sub.Particles[i].Mass;
-    msum += m;
-    for (j = 0; j < 3; j++)
-    {
-      double dx;
-      if (HBTConfig.PeriodicBoundaryOn)
-        dx = NEAREST(sub.Particles[i].ComovingPosition[j] - origin[j]);
-      else
-        dx = sub.Particles[i].ComovingPosition[j];
-      sx[j] += dx * m;
-      sx2[j] += dx * dx * m;
-    }
-  }
+    // Loop over particles in the subhalo
+    for (int i = 0; i < sub.Particles.size(); i++)
+      {
+        const int is_tracer = sub.Particles[i].IsTracer();        
+        // First pass: use tracers only
+        // Second pass: use non-tracers only
+        if((is_tracer && (pass_nr==0)) || ((!is_tracer) && (pass_nr==1))) {
 
-  for (j = 0; j < 3; j++)
+          NumPart += 1;
+          HBTReal m = sub.Particles[i].Mass;
+          msum += m;
+          for (int j = 0; j < 3; j++)
+            {
+              double dx;
+              if (HBTConfig.PeriodicBoundaryOn)
+                dx = NEAREST(sub.Particles[i].ComovingPosition[j] - origin[j]);
+              else
+                dx = sub.Particles[i].ComovingPosition[j];
+              sx[j] += dx * m;
+              sx2[j] += dx * dx * m;
+            }
+        }
+        if(NumPart==NumPartCoreMax)break;
+        // Next particle in subhalo
+      }
+        if(NumPart==NumPartCoreMax)break;
```

Comment:
Is it possible for this condition to be met if the one 3 lines above was not?

Reply:
I don't think so, but I've repeated the condition because if NumPart==NumPartCoreMax I want to break out of both the loop over i and the loop over pass_nr. That happens if we find enough particles on the first pass (where we only consider the tracer type particles) so we skip the second pass.